---
title: "Lab book: CoV data summary"
output: radix::radix_article
geometry: margin=1in, includeheadfoot
fig_captions: true
fontsize: 3pt
header-includes:
   - \usepackage{longtable}
   - \usepackage{fancyhdr}
   - \fancyfoot[LE,RO]{\thepage}
author:
  - name: Liam Brierley
    affiliation: University of Liverpool
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
---
\pagenumbering{arabic}

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, error=TRUE)
```

## Numbers of viruses, sequences

*Search terms: `r searchterm`*

- **`r n_cov`** viruses in family Coronaviridae identified through NCBI taxonomy.
- Of these **`r n_cov`**, **`r n_cov_seq`** (**`r round(n_cov_seq*100/n_cov, 1)`**%) have at least one available spike protein nucleotide sequence.
- Of these **`r n_cov_seq`**, **`r n_cov_seq_eid2`** (**`r round(n_cov_seq_eid2*100/n_cov_seq, 1)`**%) have at least one recognised mammal or bird host in EID2 (which may be human host) *NB: virtually all coronaviruses are viruses of birds/mammals*
- Total sequence accessions: **`r n_seq`**, made up of **`r n_cds`** individual protein coding sequences

```{r}
allcov_df %>% mutate(n_seqs = ifelse(n_seqs > 0, "sequence available", "no sequence"), Hostspp = ifelse(Hostspp > 0, "host recognised", "no hosts")) %>% with(table(n_seqs, Hostspp)) %>% kable
```

**Number of sequences per coronavirus heavily skewed, most just have 1:**
```{r}
allcov_df %>% filter(n_seqs > 0) %>% select(n_seqs) %>% table %>% kable
allcov_df %>% arrange(n_seqs) %>% tail(n = 5) %>% select(childtaxa_name, n_seqs) %>% kable
```

## SARS-CoV-2 sequences

- **`r allcov_df %>% filter(childtaxa_id == 2697049) %>% select(n_seqs) %>% as.integer`** spike protein nucleotide sequences identified by programmatic extraction from GenBank
- However, manual extraction from GenBank (download from [SARS-CoV-2 landing page](https://www.ncbi.nlm.nih.gov/genbank/sars-cov-2-seqs/) at 2020-03-27 18:15:04 GMT) find additional sequences not contained in these searches because metadata incorrectly labels them under taxid 694009 for SARS-CoV (the parent species).
- **`r n_man_genbank_seqs`** sequences identified from manual extraction from GenBank (not as many as above, even with additional sequences, because again, not fully annotated, so can't extract from searches)
- **`r length(sarscov2_GISAID_FASTA)`** sequences identified from manual extraction from GISAID (all complete genomes, though proteins are not annotated..)

## Numbers of hosts

Considering only the `r n_cov_seq` coronaviruses with available spike protein sequence data...

**Number of host species per coronavirus also heavily skewed, as expected:**
```{r}
allcov_df %>% filter(n_seqs > 0) %>% select(Hostspp) %>% table %>% kable
allcov_df %>% filter(n_seqs > 0) %>% arrange(Hostspp) %>% tail(n = 5) %>% select(childtaxa_name, Hostspp) %>% kable
```
*Coronaviruses with broadest host range include very wide species that encompass many individual strains..*

**Number of coronaviruses per host species also heavily skewed:**
```{r}
eid2_cov %>% filter(PathogenTaxID %in% (allcov_df %>% filter(n_seqs > 0) %>% .$childtaxa_id %>% as.integer)) %>% select(Host) %>% table %>% table %>% kable
eid2_cov %>% filter(PathogenTaxID %in% (allcov_df %>% filter(n_seqs > 0) %>% .$childtaxa_id %>% as.integer)) %>% select(Host) %>% table %>% as.data.frame %>% arrange(Freq) %>% tail(n = 6) %>% kable
```
*As expected, some commonly studied species (ferret, rat, domestic pig, human), plus livestock (alpaca) and one horseshoe bat, sequences mostly derive from a [single study](https://www.nature.com/articles/nature12711)*

**Number of coronaviruses infecting each host group:**

Host groups are mutually exclusive, i.e. primates = non-human primates. Other mammals = misc orders (Proboscidea, Eulipotyphla, Cingulata..)

```{r plot_hostcount, results="asis", fig.height=7.0, fig.width=10.0}
ggplot(data.frame(host = allcov_df %>% filter(n_seqs > 0) %>% select(starts_with("h_")) %>%
               colnames(),
           count = allcov_df %>% filter(n_seqs > 0) %>% select(starts_with("h_")) %>% 
               sapply(function(x) x %>% 
                          as.character() %>% 
                          as.numeric()) %>%
               colSums()),
       aes(x = host, y = count)) +
  geom_bar(stat="identity") +
  scale_y_continuous(expand = c(0,0), limits = c(0,40)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        text = element_text(size=20))
```

*Not too sure this is very meaningful given how little we know about potential animal hosts of coronaviruses*

## Sequence data summaries

- Total sequence accessions: **`r n_seq`**, most of which are only partial sequences for spike *(NB this distinction assigned using regexp searches in title, as most accessions are missing the "completeness" field and some appear incorrect)*

```{r}
table(allcov_meta_df$complete) %>% prop.table %>% kable
```

- While most sequences are only partial, most coronaviruses with at least 1 sequence actually have at least 1 complete spike sequence (`r n_cov_seq - (allcov_meta_df %>% group_by(taxid, complete) %>% count(complete) %>% spread(complete, n, fill=0) %>% rename_at(vars(-taxid), ~ paste0("n_",.)) %>% filter(n_complete_spike == 0 & n_whole_genome == 0) %>% nrow)` of `r n_cov_seq`), *i.e. most of these partial spikes are for the same coronaviruses*

- Trying to determine which protein coding sequences correspond to the spike proteins:
- `r round((cov_enc_df %>% filter(is.na(protein)) %>% nrow)*100/n_cds,2)`% of coding sequences do not have a protein label annotation in metadata and `r round((cov_enc_df %>% filter(is.na(gene)) %>% nrow)*100/n_cds,2)`% do not have a gene label annotation - using regular expressions based on proteins to determine whether each coding sequence is spike (may be entire S, or just S1 subunit, S2 subunit) or other:

```{r}
table(cov_enc_df$seqtype) %>% kable
```

**Excluding partial sequences**, summaries of counts of different complete coding sequence types *per coronavirus (taxid)*:

S

```{r}
table(allcov_df$n_S) %>% t %>% kable
```

S1 - *only 17 viruses have these*

```{r}
table(allcov_df$n_S1) %>% t %>% kable
```

S2 - *only 2 viruses have these (infectious bronchitis virus, porcine epidemic diarrhea virus), and they have many sequences!*

```{r}
table(allcov_df$n_S2) %>% t %>% kable
```

**Excluding partial sequences**, summaries of genomic characteristics *per coronavirus (taxid)* (i.e. values are averaged within each virus so that each virus represents only one data point):

*Only for viruses that have whole genome sequences, mean lengths:*

```{r plot_length_wgs, results="asis", layout="l-body-outset", fig.height=4.0, fig.width=8.0}
ggplot(allcov_df, aes(x = mean_wgs_length)) +  
  geom_histogram(fill="green") +
  xlab(paste0("whole genome sequence length (bases), n = ",allcov_df %>% filter(!is.na(mean_wgs_length)) %>% nrow)) +
  theme_bw()
```

*Mean lengths, ENC, GC content of spike proteins compared to other proteins:*

```{r plot_length_enc_gc, results="asis", layout="l-body-outset", fig.height=12.0, fig.width=13.0}
g2 <- ggplot(allcov_df %>% select(mean_length_other, mean_length_S) %>% melt, aes(x = value, fill = variable)) +  
  geom_density(alpha = 0.6) +
  scale_fill_discrete(name = "protein", labels = c(paste0("other (n = ",allcov_df %>% filter(!is.na(mean_length_other)) %>% nrow,")"),
                                                   paste0("spike (n = ",allcov_df %>% filter(!is.na(mean_length_S)) %>% nrow,")"))) +
  xlab("mean coding sequence length (bases)") +
  theme_bw(base_size = 14)

g3 <- ggplot(allcov_df %>% select(mean_enc_other, mean_enc_S) %>% melt, aes(x = value, fill = variable)) +  
  geom_density(alpha = 0.6) +
  scale_fill_discrete(name = "protein", labels = c(paste0("other (n = ",allcov_df %>% filter(!is.na(mean_enc_other)) %>% nrow,")"),
                                                   paste0("spike (n = ",allcov_df %>% filter(!is.na(mean_enc_S)) %>% nrow,")"))) +
  xlab("mean effective number of codons (ENC)") +
  theme_bw(base_size = 14)

g4 <- ggplot(allcov_df %>% select(mean_GC_other, mean_GC_S) %>% melt, aes(x = value*100, fill = variable)) +  
  geom_density(alpha = 0.6) +
  scale_fill_discrete(name = "protein", labels = c(paste0("other (n = ",allcov_df %>% filter(!is.na(mean_GC_other)) %>% nrow,")"),
                                                   paste0("spike (n = ",allcov_df %>% filter(!is.na(mean_GC_S)) %>% nrow,")"))) +
  xlab("mean GC content (%)") +
  theme_bw(base_size = 14)

g2 + g3 + g4 + plot_layout(ncol=1)
```

**Stronger codon bias in spike than other proteins. Reasonable variation between spike proteins, although GC content is more uniform than in other proteins!**

