---
title: "Lab book: CoV data summary"
output: radix::radix_article
geometry: margin=1in, includeheadfoot
fig_captions: true
fontsize: 3pt
header-includes:
   - \usepackage{longtable}
   - \usepackage{fancyhdr}
   - \fancyfoot[LE,RO]{\thepage}
author:
  - name: Liam Brierley
    affiliation: University of Liverpool
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
---
\pagenumbering{arabic}

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, error=TRUE)
```

## Numbers of viruses, sequences

*Search terms: `r searchterm`*

- **`r n_cov`** viruses in family Coronaviridae identified through NCBI taxonomy.
- Of these **`r n_cov`**, **`r n_cov_seq`** (**`r round(n_cov_seq*100/n_cov, 1)`**%) have at least one available spike protein nucleotide sequence.
- Of these **`r n_cov_seq`**, **`r n_cov_seq_eid2`** (**`r round(n_cov_seq_eid2*100/n_cov_seq, 1)`**%) have at least one recognised mammal or bird host in EID2 (which may be human host) *NB: virtually all coronaviruses are viruses of birds/mammals*
- Total sequence accessions: **`r n_seq`**, made up of **`r n_cds`** individual protein coding sequences

```{r}
allcov_df %>% mutate(n_seqs = ifelse(n_seqs > 0, "spike sequence available", "no spike sequence"), Hostspp = ifelse(Hostspp > 0, "host recognised", "no hosts")) %>% with(table(n_seqs, Hostspp)) %>% kable
```

**Number of sequences per coronavirus heavily skewed, most just have 1:**
```{r}
allcov_df %>% filter(n_seqs > 0) %>% select(n_seqs) %>% table %>% kable
allcov_df %>% arrange(n_seqs) %>% tail(n = 5) %>% select(childtaxa_name, n_seqs) %>% kable
```

## SARS-CoV-2 sequences

- **`r allcov_df %>% filter(childtaxa_id == 2697049) %>% select(n_seqs) %>% as.integer`** spike protein nucleotide sequences identified by programmatic extraction from GenBank
- However, manual extraction from GenBank (download from [SARS-CoV-2 landing page](https://www.ncbi.nlm.nih.gov/genbank/sars-cov-2-seqs/) at 2020-03-27 18:15:04 GMT) find additional sequences not contained in these searches because metadata incorrectly labels them under taxid 694009 for SARS-CoV (the parent species).
- **`r n_man_genbank_seqs`** sequences identified from manual extraction from GenBank (not as many as above, even with additional sequences, because again, not fully annotated, so can't extract from searches)
- **`r length(sarscov2_GISAID_FASTA)`** sequences identified from manual extraction from GISAID (all complete genomes, though proteins are not annotated..)

## Numbers of hosts

Considering only the `r n_cov_seq` coronaviruses with available spike protein sequence data...

**Number of host species per coronavirus also heavily skewed, as expected:**
```{r}
allcov_df %>% filter(n_seqs > 0) %>% select(Hostspp) %>% table %>% kable
allcov_df %>% filter(n_seqs > 0) %>% arrange(Hostspp) %>% tail(n = 5) %>% select(childtaxa_name, Hostspp) %>% kable
```
*Coronaviruses with broadest host range include very wide species that encompass many individual strains..*

**Number of coronaviruses per host species also heavily skewed:**
```{r}
eid2_cov %>% filter(PathogenTaxID %in% (allcov_df %>% filter(n_seqs > 0) %>% .$childtaxa_id %>% as.integer)) %>% select(Host) %>% table %>% table %>% kable
eid2_cov %>% filter(PathogenTaxID %in% (allcov_df %>% filter(n_seqs > 0) %>% .$childtaxa_id %>% as.integer)) %>% select(Host) %>% table %>% as.data.frame %>% arrange(Freq) %>% tail(n = 6) %>% kable
```
*As expected, some commonly studied species (ferret, rat, domestic pig, human), plus livestock (alpaca) and one horseshoe bat, sequences mostly derive from a [single study](https://www.nature.com/articles/nature12711)*

**Number of coronaviruses infecting each host group:**

Host groups are mutually exclusive, i.e. primates = non-human primates. Other mammals = misc orders (Proboscidea, Eulipotyphla, Cingulata..)

```{r plot_hostcount, results="asis", fig.height=7.0, fig.width=10.0}
ggplot(data.frame(host = allcov_df %>% filter(n_seqs > 0) %>% select(starts_with("h_")) %>%
                    colnames(),
                  count = allcov_df %>% filter(n_seqs > 0) %>% select(starts_with("h_")) %>% 
                    sapply(function(x) x %>% 
                             as.character() %>% 
                             as.numeric()) %>%
                    colSums()),
       aes(x = host, y = count)) +
  geom_bar(stat="identity") +
  scale_y_continuous(expand = c(0,0), limits = c(0,40)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        text = element_text(size=20))
```

*Not too sure this is very meaningful given how little we know about potential animal hosts of coronaviruses*

## Sequence data quality

- Total sequence accessions: **`r n_seq`**, most of which are only partial sequences for spike *(NB this distinction assigned using regexp searches in title, as most accessions are missing the "completeness" field and some appear incorrect)*

```{r}
table(allcov_meta_df$complete) %>% prop.table %>% kable
```

- While most sequences are only partial, most coronaviruses with at least 1 sequence actually have at least 1 complete spike sequence (`r allcov_df %>% filter(!is.na(n_S) & n_S > 0) %>% nrow` of `r n_cov_seq`).

- Trying to determine which protein coding sequences correspond to the spike proteins:
- `r round((cov_enc_df %>% filter(is.na(protein)) %>% nrow)*100/n_cds,2)`% of coding sequences do not have a protein label annotation in metadata and `r round((cov_enc_df %>% filter(is.na(gene)) %>% nrow)*100/n_cds,2)`% do not have a gene label annotation - using regular expressions based on proteins to determine whether each coding sequence is spike (may be entire S, or just S1 subunit, S2 subunit) or other:

```{r}
cov_enc_df %>% filter(include == "Y") %>% with(table(seqtype, complete)) %>% kable
```

**Excluding partial sequences**, summaries of counts of different complete spike protein sequence types *per coronavirus (taxid)*:

```{r}
table(allcov_df$n_S) %>% t %>% kable
```

So **`r sum(allcov_df$n_S, na.rm=TRUE)`** individual spike protein sequences across **`r allcov_df %>% filter(!is.na(n_S) & n_S > 0) %>% nrow`** viruses.

*NB further complete coding sequences are available that give subunits separately - for S1, `r allcov_df %>% filter(n_S1 > 0) %>% nrow` virus (`r allcov_df %>% filter(n_S1 > 0) %>% .$childtaxa_name`) and for S2, `r allcov_df %>% filter(n_S2 > 0) %>% nrow` virus (`r allcov_df %>% filter(n_S2 > 0) %>% .$childtaxa_name`). Not considering these for now.*

## Sequence data summaries

**Excluding partial sequences**, summaries of genomic characteristics *per coronavirus (taxid)* (i.e. values are averaged within each virus so that each virus represents only one data point):

*Only for viruses that have whole genome sequences, mean lengths:*

```{r plot_length_wgs, results="asis", layout="l-body-outset", fig.height=4.0, fig.width=8.0}
ggplot(allcov_df, aes(x = mean_wgs_length)) +  
  geom_histogram(fill="green") +
  xlab(paste0("whole genome sequence length (bases), n = ",allcov_df %>% filter(!is.na(mean_wgs_length)) %>% nrow)) +
  theme_bw()
```

*Lengths, ENC, GC content between-coronaviruses in spike and other proteins; within-coronaviruses in spike:*

```{r plot_length, results="asis", layout="l-body-outset", fig.height=9.0, fig.width=10.0}
g_len_1 <- ggplot(allcov_df %>% select(mean_length_other, mean_length_S) %>% melt, aes(x = value, fill = variable)) +  
  geom_density(alpha = 0.6) +
  scale_fill_discrete(name = "protein", labels = c(paste0("other (n = ",allcov_df %>% filter(!is.na(mean_length_other)) %>% nrow,")"),
                                                   paste0("spike (n = ",allcov_df %>% filter(!is.na(mean_length_S)) %>% nrow,")"))) +
  xlab("mean coding sequence length (bases)") +
  coord_cartesian(xlim=c(0,12000), expand=FALSE) +
  theme_bw(base_size = 14)

g_len_2 <- ggplot(cov_spikes_df %>% filter(taxid %in% (allcov_df %>% filter(n_S > 4) %>% .$childtaxa_id)) %>% mutate(taxid = abbreviate(taxid2name(taxid))), 
                  aes(x = taxid, y = length, fill = taxid)) +  
  geom_boxplot(outlier.size = 0.3) +
  guides(fill = FALSE) +
  coord_flip(ylim=c(0,12000), expand=FALSE) +
  xlab("virus (only those w/ >= 5 spike seqs)") +
  ylab("spike coding sequence length (bases)") +
  theme_bw(base_size = 14) +
  theme(panel.grid.major.y = element_blank())

g_len_1 + g_len_2 + plot_layout(ncol=1,heights=c(1,2))
```

```{r}
anova(lm(length ~ taxid, data=cov_spikes_df))
```

**Fairly consistent sequence lengths of spikes compared to other proteins (expected as pooling all others). Very little within-coronavirus variation.**  


```{r plot_enc, results="asis", layout="l-body-outset", fig.height=9.0, fig.width=10.0}
g_enc_1 <- ggplot(allcov_df %>% select(mean_enc_other, mean_enc_S) %>% melt, aes(x = value, fill = variable)) +  
  geom_density(alpha = 0.6) +
  scale_fill_discrete(name = "protein", labels = c(paste0("other (n = ",allcov_df %>% filter(!is.na(mean_enc_other)) %>% nrow,")"),
                                                   paste0("spike (n = ",allcov_df %>% filter(!is.na(mean_enc_S)) %>% nrow,")"))) +
  xlab("mean effective number of codons (ENC)") +
  coord_cartesian(xlim=c(33,60), expand=FALSE) +
  theme_bw(base_size = 14)

g_enc_2 <- ggplot(cov_spikes_df %>% filter(taxid %in% (allcov_df %>% filter(n_S > 4) %>% .$childtaxa_id)) %>% mutate(taxid = abbreviate(taxid2name(taxid))), 
                  aes(x = taxid, y = enc, fill = taxid)) +  
  geom_boxplot(outlier.size = 0.3) +
  guides(fill = FALSE) +
  coord_flip(ylim=c(33,60), expand=FALSE) +
  xlab("virus (only those w/ >= 5 spike seqs)") +
  ylab("effective number of codons (ENC)") +
  theme_bw(base_size = 14) +
  theme(panel.grid.major.y = element_blank())

g_enc_1 + g_enc_2 + plot_layout(ncol=1,heights=c(1,2))
```

```{r}
anova(lm(enc ~ taxid, data=cov_spikes_df))
```

**Stronger average codon bias in spike than other proteins! Reasonable variation in spike codon biases between-coronaviruses and within some coronaviruses. Human CoV HKU1 more strongly biased than other CoVs.**  


```{r plot_gc, results="asis", layout="l-body-outset", fig.height=9.0, fig.width=10.0}
g_gc_1 <- ggplot(allcov_df %>% select(mean_GC_other, mean_GC_S) %>% melt, aes(x = value, fill = variable)) +  
  geom_density(alpha = 0.6) +
  scale_fill_discrete(name = "protein", labels = c(paste0("other (n = ",allcov_df %>% filter(!is.na(mean_GC_other)) %>% nrow,")"),
                                                   paste0("spike (n = ",allcov_df %>% filter(!is.na(mean_GC_S)) %>% nrow,")"))) +
  xlab("mean GC content (%)") +
  coord_cartesian(xlim=c(30,48), expand=FALSE) +
  theme_bw(base_size = 14)

g_gc_2 <- ggplot(cov_spikes_df %>% filter(taxid %in% (allcov_df %>% filter(n_S > 4) %>% .$childtaxa_id)) %>% mutate(taxid = abbreviate(taxid2name(taxid))), 
                 aes(x = taxid, y = GC_content, fill = taxid)) +  
  geom_boxplot(outlier.size = 0.3) +
  guides(fill = FALSE) +
  coord_flip(ylim=c(30,48), expand=FALSE) +
  xlab("virus (only those w/ >= 5 spike seqs)") +
  ylab("GC content (%)") +
  theme_bw(base_size = 14) +
  theme(panel.grid.major.y = element_blank())

g_gc_1 + g_gc_2 + plot_layout(ncol=1,heights=c(1,2))
```

```{r}
anova(lm(G+C ~ taxid, data=cov_spikes_df))
```

**GC content slightly lower and slightly more uniform in spikes than in other proteins! Some variation between-coronaviruses, and some variation within-coronaviruses. Human CoV HKU1 and Wencheng shrew Cov more strongly biased than other CoVs.**  


**Mean GC content of spike versus known host range count, labelled as human/nonhuman virus**
```{r plot_GC_hostrange, results="asis", layout="l-screen-inset", fig.height=8.0, fig.width=8.0}
g <- ggplot(allcov_df,
            aes(x=mean_GC_S, y=log(Hostspp+1), fill=h_human, label=childtaxa_name, label2=Hostspp)) +
  geom_point(aes(color=h_human)) +
  theme_bw()

ggplotly(g)
```
*Not too informative though useful to see which virus is which?*

## Spike protein composition

```{r plot_dinuc_all_genus, results="asis", layout="l-screen-inset", fig.height=7.0, fig.width=10.0}
ggplot(merge(cov_spikes_df,
             allcov_df %>% select(childtaxa_id, genus) %>% mutate(genus = gsub("0","unclassified",genus)),
             all.x = TRUE,
             by.x = "taxid",
             by.y = "childtaxa_id") %>% 
         select(genus, matches("^[A|C|G|T][A|C|G|T]_Bias$")) %>% melt(id.vars = "genus"), 
       aes(y = value, x = genus, fill = genus)) +  
  geom_boxplot(outlier.size=0.3) +
  facet_wrap(. ~ variable) +
  theme_bw(base_size = 11) +
  theme(axis.text.x=element_blank()) +
  geom_hline(yintercept=1, alpha = 0.5, color="grey50", lty="dashed", size=1) +
  xlab("Genus") +
  ylab("Dinucleotide bias")
```
*Dinucleotide biases do vary in scale - clearly some biases present (TG overrepresented, CG underepresented). But these are pretty consistent between genera.*
  
  
```{r plot_dinuc_pos, results="asis", layout="l-screen-inset", fig.height=7.0, fig.width=12.0}
ggplot(cov_spikes_df %>% select(matches("^[A|C|G|T][A|C|G|T]_p[1|2|3]_Bias$")) %>% melt %>% 
    mutate(pos = substr(variable,4,5), 
           pos = case_when(pos == "p1" ~ "1-2", pos == "p2" ~ "2-3", pos == "p3" ~ "3-1"),
           variable = paste0(substr(variable,1,2),"_Bias")), 
       aes(y = value, x = pos, fill = pos)) +  
  geom_boxplot(outlier.size=0.3) +
  facet_wrap(. ~ variable) +
  theme_bw(base_size = 11) +
  theme(axis.text.x=element_blank()) +
  geom_hline(yintercept=1, alpha = 0.5, color="grey50", lty="dashed", size=1) +
  xlab("Position") +
  ylab("Dinucleotide bias")
```
*Reassuring - biases are more extreme at bridge (3-1) dinucleotides as expected. TG, TA, CA overrepresented, GT, GA, AT underepresented. Still sufficient variability to look for signal in*
  
  
```{r plot_codon_genus, results="asis", layout="l-screen-inset", fig.height=7.0, fig.width=12.0}
ggplot(merge(cov_spikes_df,
             allcov_df %>% select(childtaxa_id, genus) %>% mutate(genus = gsub("0","unclassified",genus)),
             all.x = TRUE,
             by.x = "taxid",
             by.y = "childtaxa_id") %>% 
         select(genus, matches("^[A|C|G|T][A|C|G|T][A|C|G|T]_Bias$")) %>% melt(id.vars = "genus"), 
       aes(y = value, x = genus, fill = genus)) +  
  geom_boxplot(outlier.size=0.3) +
  facet_wrap(. ~ variable) +
  theme_bw(base_size = 8) +
  theme(axis.text.x=element_blank()) +
  geom_hline(yintercept=1, alpha = 0.5, color="grey50", lty="dashed", size=1) +
  xlab("Genus") +
  ylab("Codon bias (RSCU)")
```
*Most obvious thing is use of different stop codons. But otherwise, fairly consistent across genera agaih..*  
  

```{r plot_aacid_genus, results="asis", layout="l-screen-inset", fig.height=7.0, fig.width=12.0}
ggplot(merge(cov_spikes_df,
             allcov_df %>% select(childtaxa_id, childtaxa_name, genus) %>% mutate(genus = gsub("0","unclassified",genus)),
             all.x = TRUE,
             by.x = "taxid",
             by.y = "childtaxa_id") %>% 
         select(childtaxa_name, genus, matches("^.*_aa_Bias$")) %>% melt(id.vars = c("childtaxa_name", "genus")), 
       aes(y = value, x = genus, fill = genus, label = childtaxa_name)) +  
  #  geom_point(alpha = 0) +
  geom_boxplot(outlier.size=0.3) +
  facet_wrap(. ~ variable, scales = "free_y") +
  theme_bw(base_size = 11) +
  theme(axis.text.x=element_blank()) +
#  geom_hline(yintercept=1/21, alpha = 0.5, color="grey50", lty="dashed", size=1) +
  xlab("Genus") +
  ylab("Amino acid bias")
```
*Not convinced amino acid bias is really useful here - it's just proportion amino acids in the protein sequence, and it'll be fairly consistent between CoVs..*  
  