---
title: "Lab book: PCA outputs, whole genomes"
output: radix::radix_article
geometry: margin=1in, includeheadfoot
fig_captions: true
fontsize: 3pt
header-includes:
   - \usepackage{longtable}
   - \usepackage{fancyhdr}
   - \fancyfoot[LE,RO]{\thepage}
author:
  - name: Liam Brierley
    affiliation: University of Liverpool
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
---
\pagenumbering{arabic}

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, error=TRUE)
```

**PCAs based on genome composition for `r cov_wg_df %>% nrow` coronavirus whole genome sequences. PCAs are colour-coded and ellipses drawn based on different outcome variables, though underlying PCA for each bias type is the same. Mouseover gives outcome variable and virus name.**

## Dinucleotide bias // genus

```{r plot_biplot_dinucs_genus, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "dinucs", outcome = "genus")
```

_As for spikes, strong separation, but scattered betacovs. Outlying cetacean viruses on left??_

## Dinucleotide bias // human infection capability

```{r plot_biplot_dinucs_human, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "dinucs", outcome = "h_human")
```

_As for spikes, tight clusters of individual viruses._

## Codon bias (RSCU) // genus

```{r plot_biplot_codons_genus, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "codons", outcome = "genus")
```

_Stronger separation of genera than spikes, likely because stop codons are not so strongly weighted. Alphacovs still seem to prefer TGA, but not so clear for other genera._

## Codon bias (RSCU) // human infection capability

```{r plot_biplot_codons_human, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "codons", outcome = "h_human")
```

_Epidemic coronaviruses fairly distant from endemic coronaviruses here, unlike spikes. Not as strong a pattern of preference for stop codons for all human coronaviruses either (unlike spikes). Not worth looking at without stop codons as they're not overly influential here._

## Codon bias (RSCU) without stop codon // human infection capability // PCA3 and PCA4

```{r plot_biplot_codonnostop_human_extra, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "codons_nostop", outcome = "h_human", choices=3:4)
```

_**However** if excluding stop codons, similar pattern emerges to spikes when examining PC3/4, even though they explain small amount of variance, good separation on human vs nonhuman viruses. Real signal or just noise...??_

## Amino acid bias // genus

```{r plot_biplot_aa_genus, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "aa", outcome = "genus")
```

_Unlike spikes, very strong separation when PCA accounts for whole genome._

## Amino acid bias // human infection capability

```{r plot_biplot_aa_human, results="asis", layout="l-body-outset", fig.height=10.0, fig.width=10.0}
genomic_pca(cov_wg_df, vars = "aa", outcome = "h_human")
```

_As for spikes, reasonable separation between epidemic coronaviruses and endemic coronaviruses!_